libraries/freertos_plus/standard/freertos_plus_tcp/source/FreeRTOS_DNS.c

```diff
@@ -451,7 +503,9 @@ TickType_t xIdentifier = 0;
 	/* Generate a unique identifier. */
 	if( 0 == ulIPAddress )
 	{
-		xIdentifier = ( TickType_t )ipconfigRAND32( );
+		/* DNS identifiers are 16-bit. */
+		uxIdentifier = ( TickType_t ) ( ipconfigRAND32() & 0xffffu );
+		/* ipconfigRAND32() may not return a non-zero value. */
```

Comment:
Earlier I had a long discussion with @dcgaws about `ipconfigRAND32( )`, but without a conclusion yet.
The macro is supposed to return a 32-bit random number. It should also have a random initial value ( c.q. seed ).
Some random generators will return 0 when they are broken ( not able to start up ). In several places within +TCP you will see code that checks for a "valid" random value, where "valid" means: non-zero.
dcgaws argued that 0 is also a valid random number.

    /* ipconfigRAND32() may not return a non-zero value. */

The comment means: "`ipconfigRAND32()` shall not return zero".

What to do? There is a lot of code that depends on the macro, so we can not change a lot.

In my own projects, `ipconfigRAND32()` returns non-zero values only:

    uint32_t ulRAND32()
    {
    uint32_t ulResult;
        do
        {
            ulResult = randomise();
        } while ( ulResult == 0uL );
        return ulResult;
    }

Reply:
I think one retry on a 0 return seems like a good idea.  If you get two zeroes in a row, that seems like a strong indication of an error.  Spinning forever doesn't seem to follow the principle of least astonishment.  Is there a way we can degrade gracefully?

Reply:
I was aware that the above function may end up spinning for ever. And in that way it would work like an ASSERT.
You're right that 2 zero's on a row are really rare for a 32-bit randomiser.
The above is the responsibility of the application, and I think that at most we can give advise.

Reply:
I talked this over with Richard and Cobus today.   Since there's only half a dozen uses of `ipconfigRAND32` Richard thinks we should change this interface so that errors are explicit.  Would you add a wrapper for the macro for use here and in other +TCP sources, then modify this code to treat a failed rand call as a failed DNS lookup.  I'll pursue the "degrade gracefully" idea independently.

Reply:
@dan4thewin : that is the best solution! No more doubts about the zero value,

I will propose a new function, please shoot:

```c
#ifdef ipconfigRAND32
    #pragma warning "The macro ipconfigRAND32 has been replaced with the function ulRandom32()"
#endif

extern BaseType_t ulRandom32( uint32_t *pulRandomValue );

/* For example: */

BaseType_t ulRandom32( uint32_t *pulRandomValue )
{
uint32_t ulValue;

    configASSERT( pulRandomValue != NULL );

    /* Get 32 random bits from a 15-bit rand() function. */
    *( pulRandomValue ) =
        ( ( ( ( uint32_t ) rand() ) & 0x7fffuL ) << 17 ) |
        ( ( ( ( uint32_t ) rand() ) & 0x7fffuL ) <<  2 ) |
        ( ( ( ( uint32_t ) rand() ) & 0x0003uL ) );

    /* Return pdPASS when the randomiser works OK.  When using a
    hardware randomiser which fails, pdFAIL shall be returned. */
    return pdPASS;
}
```

Reply:
I'm intrigued.  I was expecting to see ipconfigRAND32 called inside the new function.  I see in your comment you say pdFAIL is returned when the hardware randomiser fails, but I don't see how this code would return that value.  I must be missing something.

Reply:
Agreed, we need to use our NIST-grade RNG in all cases, and not the C runtime primitive.

Reply:
To @dan4thewin 
`BaseType_t xRandom32( uint32_t *pulValue )`
The random number will be passed in the parameter `uint32_t *pulValue`.

The existing `ipconfigRAND32()` macro does not tell if the result is reliable, and so it is useless.

Yesterday I wrote this function for an STM32Fx:

```c
BaseType_t xRandom32( uint32_t *pulValue )
{
HAL_StatusTypeDef xResult;
HAL_StatusTypeDef xReturn;
uint32_t ulValue;

    xResult = HAL_RNG_GenerateRandomNumber( &hrng, &ulValue );
    if( xResult == HAL_OK )
    {
        xReturn = pdPASS;
        *pulValue = ulValue;
    }
    else
    {
        xReturn = pdFAIL;
    }
    return xReturn;
}
```
because I want a random local-link IP-address at start-up. When `xRandom32()` fails, the device can not work.

Reply:
Hein, hello, we've got an implementation of a cryptographic RNG on every port. It's whatever ipconfigRAND32 is defined as. Where possible, the latter uses the capability of the underlying hardware. An example is ulRand. It does the right thing cryptographically (and handles thread safety, which is required by callers of that routine) but doesn't return an explicit status code. The latter is what would need to be changed and then percolated through the code.